### PR TITLE
[ACS-8202] basic flow getting ai response for one or more selected files

### PR DESCRIPTION
**Pull request by AleksanderSklorz**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [ ] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [x] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)
https://hyland.atlassian.net/browse/ACS-8202

**What is the new behaviour?**
Possibility to run Knowledge Retrieval is added. Implemented results page for Knowledge Retrieval (for selected files). 

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
Unit tests will be added as part of https://hyland.atlassian.net/browse/ACS-8398
Backend integration will be added whenever backend will be ready as part of https://hyland.atlassian.net/browse/ACS-8399
ADF PR: https://github.com/Alfresco/alfresco-ng2-components/pull/9944

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -63,7 +63,9 @@
     "pdfjs-dist": "3.3.122",
     "rxjs": "6.6.6",
     "tslib": "^2.5.3",
-    "zone.js": "0.11.8"
+    "zone.js": "0.11.8",
+    "lottie-web": "^5.12.2",
+    "ngx-lottie": "^9.1.0"
```
Comment (DenysVuika) [inline package.json:68]: it it mandatory to use this library? it seems they also have some analytics:

Please note: ngx-lottie uses [Scarf](https://scarf.sh/) to collect anonymized installation analytics. These analytics help support the maintainers of this library. However, if you'd like to opt out, you can do so by setting scarfSettings.enabled = false in your project's package.json. Alternatively, you can set the environment variable SCARF_ANALYTICS=false before you install.

File: projects/aca-content/assets/images/ai-stars-icon.json
Comment (DenysVuika) [inline projects/aca-content/assets/images/ai-stars-icon.json:1]: is this file required? it looks like a json, not SVG

File: projects/aca-content/src/lib/aca-content.module.ts [hunk 2]
```diff
@@ -97,7 +99,10 @@ import { SearchResultsRowComponent } from './components/search/search-results-ro
     AcaFolderRulesModule,
     CreateFromTemplateDialogComponent,
     OpenInAppComponent,
-    UploadFilesDialogComponent
+    UploadFilesDialogComponent,
+    LottieModule.forRoot({
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/aca-content.module.ts:103]: this might cause lazy loading problems, as 'aca-content' is a library module, not an application one

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/agents-button/agents-button.component.scss [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    &-agent {
+      height: 40px;
+
+      &-avatar {
+        vertical-align: middle;
+        margin-bottom: 2px;
+        margin-right: 12px;
+        border-radius: 50%;
+        background-color: #9acd9a; // temporary to keep place for avatar, will be replaced with avatar image during synchronization with backend
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/agents-button/agents-button.component.scss:41]: we now have the `adf-avatar` component that shows image or fallback to initials, maybe it can be used in this feature

File: projects/aca-content/src/lib/components/files/files.component.ts [hunk 4]
```diff
@@ -64,23 +77,41 @@ import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
...
   queryParams = null;
   showLoader$ = this.store.select(showLoaderSelector);
-  private nodePath: PathElement[];
-
   columns: DocumentListPresetRef[] = [];
   isFilterHeaderActive = false;
 
-  constructor(private contentApi: ContentApiService, private nodeActionsService: NodeActionsService, private route: ActivatedRoute) {
+  private nodePath: PathElement[];
+  private _searchAiInputState: SearchAiInputState = {
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/components/files/files.component.ts:101]: maybe worth putting that into the base `PageComponent` not to copy-paste for each page file

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import { forkJoin } from 'rxjs';
+
+@Component({
+  standalone: true,
+  imports: [
+    CommonModule,
+    PageLayoutComponent,
+    ToolbarActionComponent,
+    ToolbarModule,
+    MaterialModule,
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts:45]: the `MaterialModule` should not be used anywhere

File: projects/aca-content/src/lib/components/recent-files/recent-files.component.ts [hunk 6]
```diff
@@ -54,19 +55,35 @@ import { TranslateModule } from '@ngx-translate/core';
...
   templateUrl: './recent-files.component.html',
-  encapsulation: ViewEncapsulation.None
+  styleUrls: ['./recent-files.component.scss'],
+  encapsulation: ViewEncapsulation.None,
+  selector: 'aca-recent-files'
 })
 export class RecentFilesComponent extends PageComponent implements OnInit {
   columns: DocumentListPresetRef[] = [];
 
+  private _searchAiInputState: SearchAiInputState = {
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/components/recent-files/recent-files.component.ts:70]: repetitive code, can be placed in the base class

File: projects/aca-content/src/lib/components/recent-files/recent-files.component.scss [hunk 7]
```diff
@@ -0,0 +1,9 @@
+aca-recent-files {
+  .aca-header-container {
```
Comment (DenysVuika) [inline projects/aca-content/src/lib/components/recent-files/recent-files.component.scss:2]: this is a repetitive class, could be placed into a single place somewhere

File: projects/aca-content/assets/i18n/en.json [hunk 8]
```diff
@@ -603,5 +603,30 @@
             "BOOKS-24PX": "file library",
             "BASELINE-LOCK-24PX": "locked file"
         }
+    },
+    "KNOWLEDGE_RETRIEVAL": {
+      "SEARCH": {
+        "RESULTS_PAGE": {
+          "QUERY_INPUT_PLACEHOLDER": "Would you like to ask anything else?",
+          "REFERENCED_DOCUMENTS_HEADER": "Referenced documents",
+          "LOADING_ERROR": "Hmm...something seems to have gone wrong.",
```
Comment (MichalKinas) [inline projects/aca-content/assets/i18n/en.json:612]: Maybe API will return something more specific? With that users might be quite confused

File: projects/aca-content/assets/images/ai-stars-icon.json
Comment (MichalKinas) [reply projects/aca-content/assets/images/ai-stars-icon.json:1]: I think it is required by a library that was used to describe the animated icon

File: package.json (hunk 1 above)
Comment (AleksanderSklorz) [reply package.json:68]: I'm waiting for information about that. Will let you know when I get answer.

File: projects/aca-content/assets/images/ai-stars-icon.json
Comment (AleksanderSklorz) [reply projects/aca-content/assets/images/ai-stars-icon.json:1]: Yes, exactly. That lottie library requires describing animation in json as Michał said.

File: projects/aca-content/src/lib/aca-content.module.ts (hunk 2 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/aca-content.module.ts]: I have changed it a bit now

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/agents-button/agents-button.component.scss (hunk 3 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/agents-button/agents-button.component.scss]: replaced with adf-avatar now

File: projects/aca-content/src/lib/components/files/files.component.ts (hunk 4 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/files/files.component.ts]: good point, thanks, done

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts (hunk 5 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts]: done

File: projects/aca-content/src/lib/components/recent-files/recent-files.component.ts (hunk 6 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/recent-files/recent-files.component.ts]: done

File: projects/aca-content/src/lib/components/recent-files/recent-files.component.scss (hunk 7 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/recent-files/recent-files.component.scss]: done

File: projects/aca-content/assets/i18n/en.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/assets/i18n/en.json]: done

File: projects/aca-content/src/lib/store/effects/search-ai.effects.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+@Injectable()
+export class SearchAiEffects {
+  constructor(private actions$: Actions, private searchNavigationService: SearchAiNavigationService, private searchAiService: SearchAiService) {}
+
+  searchByTerm$ = createEffect(
+    () =>
+      this.actions$.pipe(
+        ofType<SearchByTermAiAction>(SearchAiActionTypes.SearchByTermAi),
+        map((action) => {
+          const query = action.payload.searchTerm.replace(/[(]/g, '%28').replace(/[)]/g, '%29');
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/store/effects/search-ai.effects.ts]: done

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    super();
+  }
+
+  ngOnInit(): void {
+    this.route.queryParams.pipe(takeUntil(this.onDestroy$)).subscribe((params: Params) => {
+      this._agentId = params.agentId;
+      this._searchQuery = params.query ? decodeURIComponent(params.query) : '';
+      this.restrictionQuery = params.restrictionQuery || '';
+      if (this.searchQuery) {
+        this.performAiSearch();
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts]: I changed it a bit. Now I'm exiting earlier from that function. I did not use switchMap here because performAiSearch as subscribe inside so it's not returning Observable to use switchMap. But it should work fine, I tested it and I did not see any issues.

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+  performAiSearch(): void {
+    this._loading = true;
+    this.searchAiService
+      .ask({
+        question: this.searchQuery,
+        restrictionQuery: this.restrictionQuery
+      })
+      .pipe(
+        concatMap((response) => this.searchAiService.getAnswer(response.questionId)),
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts]: changed to switchMap

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    super.ngOnInit();
+  }
+
+  ngOnDestroy(): void {
+    this.onDestroy$.next();
+    this.onDestroy$.complete();
+  }
+
+  copyResponseToClipboard(): void {
+    this.clipboardService.copyContentToClipboard(this.queryAnswer.answer, 'Copied response to clipboard');
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts]: done

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+  private _nodes: Node[] = [];
+  private restrictionQuery = '';
+  private _searchQuery = '';
+  private _queryAnswer: AiAnswer;
+
+  get agentId(): string {
+    return this._agentId;
+  }
+
+  get error(): boolean {
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-results/search-ai-results.component.ts]: done

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.ts [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    this.onDestroy$.next();
+    this.onDestroy$.complete();
+  }
+
+  onSearchSubmit(): void {
+    const error = this.searchAiService.checkSearchAvailability(this.selectedNodesState);
+    if (error) {
+      this.notificationService.showInfo(error);
+    } else {
+      const payload = new AiSearchByTermPayload();
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.ts]: now it's interface so I changed it a bit

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.scss [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+  .aca-search-ai-input-agent-select {
+    width: 149px;
+    height: 35px;
+    align-content: center;
+    border-radius: 16px;
+    padding-left: 3px;
+    padding-right: 10px;
+    background-color: var(--theme-grey-text-background-color);
+    color: var(--theme-text-light-color);
+    font-size: 15.42px;
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.scss:43]: I took all values from figma, for example for 15.42px: 
image

image

So I would keep it to be consistent with design

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input-container/search-ai-input-container.component.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  placeholder = 'KNOWLEDGE_RETRIEVAL.SEARCH.SEARCH_INPUT.DEFAULT_PLACEHOLDER';
+  @Input()
+  agentId: string;
+  @Input()
+  useStoredNodes: boolean;
+
+  constructor(private searchAiService: SearchAiService, private searchNavigationService: SearchAiNavigationService, private router: Router) {}
+
+  ngOnInit(): void {
+    this.router.events.pipe(filter((event) => event instanceof NavigationStart)).subscribe(() => this.hideSearchInput());
```
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input-container/search-ai-input-container.component.ts]: done

File: projects/aca-content/assets/i18n/en.json (hunk 8 above)
Comment (AleksanderSklorz) [reply projects/aca-content/assets/i18n/en.json]: For now we have not integrated backend. Everything are mocks. If backend will return anything helpful we will change it during integration

File: projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.scss (hunk 15 above)
[the remark this answers is not in the document]
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/components/knowledge-retrieval/search-ai/search-ai-input/search-ai-input.component.scss]: I checked it how it looks with rounded values and I can't see difference so I changed it to rounded value

File: package.json (hunk 1 above)
Comment (AleksanderSklorz) [reply package.json]: after discussion with designers I have removed that library

File: projects/aca-content/assets/images/ai-stars-icon.json
Comment (AleksanderSklorz) [reply projects/aca-content/assets/images/ai-stars-icon.json:1]: I have removed that json as we don't need animation anymore

File: projects/aca-content/src/lib/aca-content.module.ts (hunk 2 above)
Comment (AleksanderSklorz) [reply projects/aca-content/src/lib/aca-content.module.ts]: removed lottie library
